File: TowerDefense/v2/TowerDefense.py
import tkinter as tk
import math

WIDTH, HEIGHT = 850, 620
PATH = [(0,310), (220,310), (220,130), (540,130), (540,440), (850,440)]

class Enemy:
    def __init__(self, wave):
        self.x, self.y = PATH[0]
        self.target = 1
        self.health = 3 + wave
        self.max_health = self.health
        self.speed = 1.25 + wave * 0.09
        self.finished = False

    def move(self):
        if self.target >= len(PATH):
            self.finished = True
            return
        tx, ty = PATH[self.target]
        dx, dy = tx - self.x, ty - self.y
        dist = math.hypot(dx, dy)
        if dist < self.speed:
            self.x, self.y = tx, ty
            self.target += 1
        else:
            self.x += self.speed * dx / dist
            self.y += self.speed * dy / dist

class Tower:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.range = 125
        self.cooldown = 0

class Shot:
    def __init__(self, x, y, target):
        self.x = x
        self.y = y
        self.target = target
        self.speed = 9
        self.hit = False

    def move(self):
        dx, dy = self.target.x - self.x, self.target.y - self.y
        dist = math.hypot(dx, dy)
        if dist < self.speed:
            self.hit = True
        else:
            self.x += self.speed * dx / dist
            self.y += self.speed * dy / dist

class EnhancedTowerDefense:
    def __init__(self):
        self.root = tk.Tk()
        self.root.title("Enhanced Python Tower Defense")
        self.canvas = tk.Canvas(self.root, width=WIDTH, height=HEIGHT, bg="#19502d")
        self.canvas.pack()

        self.enemies = []
        self.towers = []
        self.shots = []
        self.money = 125
        self.lives = 12
        self.score = 0
        self.wave = 1
        self.spawn_timer = 0
        self.game_over = False

        self.canvas.bind("<Button-1>", self.place)
        self.loop()
        self.root.mainloop()

    def place(self, e):
        if not self.game_over and self.money >= 25:
            self.towers.append(Tower(e.x, e.y))
            self.money -= 25

    def update(self):
        if self.game_over:
            return
        self.spawn_timer += 1
        if self.spawn_timer > max(22, 82 - self.wave * 5):
            self.enemies.append(Enemy(self.wave))
            self.spawn_timer = 0

        for enemy in self.enemies[:]:
            enemy.move()
            if enemy.finished:
                self.lives -= 1
                self.enemies.remove(enemy)
            elif enemy.health <= 0:
                self.money += 10
                self.score += 10
                if self.score % 120 == 0:
                    self.wave += 1
                self.enemies.remove(enemy)

        for tower in self.towers:
            tower.cooldown -= 1
            if tower.cooldown <= 0:
                for enemy in self.enemies:
                    if math.hypot(tower.x-enemy.x, tower.y-enemy.y) <= tower.range:
                        self.shots.append(Shot(tower.x, tower.y, enemy))
                        tower.cooldown = 30
                        break

        for shot in self.shots[:]:
            shot.move()
            if shot.hit:
                shot.target.health -= 1
                self.shots.remove(shot)

        if self.lives <= 0:
            self.game_over = True

    def draw(self):
        self.canvas.delete("all")
        self.canvas.create_rectangle(0, 0, WIDTH, HEIGHT, fill="#19502d", outline="")
        for i in range(25):
            x = (i * 83) % WIDTH
            y = 65 + (i * 47) % 500
            self.canvas.create_oval(x, y, x+20, y+12, fill="#144123", outline="")

        for i in range(len(PATH)-1):
            self.canvas.create_line(PATH[i], PATH[i+1], fill="#5f462d", width=58, capstyle=tk.ROUND, joinstyle=tk.ROUND)
        for i in range(len(PATH)-1):
            self.canvas.create_line(PATH[i], PATH[i+1], fill="#916e46", width=40, capstyle=tk.ROUND, joinstyle=tk.ROUND)

        for tower in self.towers:
            self.canvas.create_oval(tower.x-tower.range, tower.y-tower.range, tower.x+tower.range, tower.y+tower.range, outline="#ffffff")
            self.canvas.create_oval(tower.x-22, tower.y-22, tower.x+22, tower.y+22, fill="#1e2855", outline="white")
            self.canvas.create_oval(tower.x-16, tower.y-16, tower.x+16, tower.y+16, fill="#4b6ed2", outline="")
            self.canvas.create_rectangle(tower.x-5, tower.y-34, tower.x+5, tower.y-2, fill="#333333", outline="")

        for enemy in self.enemies:
            self.canvas.create_oval(enemy.x-15, enemy.y+9, enemy.x+15, enemy.y+19, fill="#000000", outline="")
            self.canvas.create_oval(enemy.x-14, enemy.y-14, enemy.x+14, enemy.y+14, fill="#a52d2d", outline="")
            self.canvas.create_oval(enemy.x-7, enemy.y-8, enemy.x-2, enemy.y-3, fill="#ffa0a0", outline="")
            self.canvas.create_oval(enemy.x+3, enemy.y-8, enemy.x+8, enemy.y-3, fill="#ffa0a0", outline="")
            self.canvas.create_rectangle(enemy.x-18, enemy.y-25, enemy.x+18, enemy.y-20, fill="black", outline="")
            self.canvas.create_rectangle(enemy.x-18, enemy.y-25, enemy.x-18 + 36*max(0, enemy.health)/enemy.max_health, enemy.y-20, fill="green", outline="")

        for shot in self.shots:
            self.canvas.create_oval(shot.x-5, shot.y-5, shot.x+5, shot.y+5, fill="yellow", outline="white")

        self.canvas.create_rectangle(10, 10, 830, 52, fill="#000000", outline="#444444")
        self.canvas.create_text(330, 32, text=f"Money: ${self.money}   Lives: {self.lives}   Score: {self.score}   Wave: {self.wave}   Click to place cannon tower ($25)",
                                fill="white", font=("Arial", 16, "bold"))

        if self.game_over:
            self.canvas.create_rectangle(0, 250, WIDTH, 350, fill="#000000")
            self.canvas.create_text(WIDTH//2, 310, text="GAME OVER", fill="red", font=("Arial", 42, "bold"))

    def loop(self):
        self.update()
        self.draw()
        self.root.after(16, self.loop)

EnhancedTowerDefense()

File: TowerDefense/v2/TowerDefense.java
import java.awt.*;
import java.awt.event.*;
import java.util.ArrayList;
import javax.swing.*;

public class TowerDefense extends JFrame {
    public TowerDefense() {
        add(new TDPanel());
        setTitle("Tower Defense");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);
        pack();
        setLocationRelativeTo(null);
        setVisible(true);
    }

    public static void main(String[] args) {
        new TowerDefense();
    }
}

class TDPanel extends JPanel implements ActionListener, MouseListener {
    private final int WIDTH = 850;
    private final int HEIGHT = 620;
    private Timer timer = new Timer(16, this);

    private ArrayList<TDEnemy> enemies = new ArrayList<>();
    private ArrayList<TDTower> towers = new ArrayList<>();
    private ArrayList<TDShot> shots = new ArrayList<>();
    private Point[] path = {
        new Point(0, 310), new Point(220, 310), new Point(220, 130),
        new Point(540, 130), new Point(540, 440), new Point(850, 440)
    };

    private int money = 125, lives = 12, score = 0, wave = 1, spawnTimer = 0, anim = 0;
    private boolean gameOver = false;

    public TDPanel() {
        setPreferredSize(new Dimension(WIDTH, HEIGHT));
        addMouseListener(this);
        timer.start();
    }

    protected void paintComponent(Graphics g) {
        Graphics2D g2 = (Graphics2D)g;
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        g2.setPaint(new GradientPaint(0, 0, new Color(25, 80, 45), 0, HEIGHT, new Color(12, 45, 35)));
        g2.fillRect(0, 0, WIDTH, HEIGHT);

        drawTerrain(g2);
        drawPath(g2);

        for (TDTower t : towers) t.draw(g2);
        for (TDEnemy e : enemies) e.draw(g2);
        for (TDShot s : shots) s.draw(g2);

        drawHUD(g2);

        if (gameOver) {
            g2.setColor(new Color(0,0,0,180));
            g2.fillRect(0, 250, WIDTH, 100);
            g2.setColor(Color.RED);
            g2.setFont(new Font("Arial", Font.BOLD, 44));
            g2.drawString("GAME OVER", 290, 315);
        }
    }

    private void drawTerrain(Graphics2D g2) {
        g2.setColor(new Color(20, 65, 35));
        for (int i = 0; i < 25; i++) {
            int x = (i * 83) % WIDTH;
            int y = 65 + (i * 47) % 500;
            g2.fillOval(x, y, 20, 12);
        }
    }

    private void drawPath(Graphics2D g2) {
        g2.setStroke(new BasicStroke(58, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND));
        g2.setColor(new Color(95, 70, 45));
        for (int i = 0; i < path.length - 1; i++) g2.drawLine(path[i].x, path[i].y, path[i+1].x, path[i+1].y);

        g2.setStroke(new BasicStroke(40, BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND));
        g2.setColor(new Color(145, 110, 70));
        for (int i = 0; i < path.length - 1; i++) g2.drawLine(path[i].x, path[i].y, path[i+1].x, path[i+1].y);

        g2.setStroke(new BasicStroke(1));
    }

    private void drawHUD(Graphics2D g2) {
        g2.setColor(new Color(0,0,0,170));
        g2.fillRoundRect(10, 10, 820, 42, 15, 15);
        g2.setColor(Color.WHITE);
        g2.setFont(new Font("Arial", Font.BOLD, 18));
        g2.drawString("Money: $" + money, 25, 37);
        g2.drawString("Lives: " + lives, 165, 37);
        g2.drawString("Score: " + score, 275, 37);
        g2.drawString("Wave: " + wave, 390, 37);
        g2.drawString("Click to place cannon tower ($25)", 510, 37);
    }

    public void actionPerformed(ActionEvent e) {
        anim++;
        if (!gameOver) {
            spawn();
            updateEnemies();
            updateTowers();
            updateShots();
            if (lives <= 0) gameOver = true;
        }
        repaint();
    }

    private void spawn() {
        spawnTimer++;
        if (spawnTimer > Math.max(22, 82 - wave * 5)) {
            enemies.add(new TDEnemy(path, wave));
            spawnTimer = 0;
        }
    }

    private void updateEnemies() {
        for (int i = enemies.size()-1; i >= 0; i--) {
            TDEnemy enemy = enemies.get(i);
            enemy.move();
            if (enemy.finished) {
                lives--;
                enemies.remove(i);
            } else if (enemy.health <= 0) {
                money += 10;
                score += 10;
                if (score % 120 == 0) wave++;
                enemies.remove(i);
            }
        }
    }

    private void updateTowers() {
        for (TDTower t : towers) {
            t.cooldown--;
            if (t.cooldown <= 0) {
                TDEnemy target = null;
                for (TDEnemy enemy : enemies) {
                    if (Math.hypot(t.x - enemy.x, t.y - enemy.y) <= t.range) {
                        target = enemy;
                        break;
                    }
                }
                if (target != null) {
                    shots.add(new TDShot(t.x, t.y, target));
                    t.cooldown = 30;
                }
            }
        }
    }

    private void updateShots() {
        for (int i = shots.size()-1; i >= 0; i--) {
            TDShot s = shots.get(i);
            s.move();
            if (s.hit) {
                s.target.health--;
                shots.remove(i);
            }
        }
    }

    public void mousePressed(MouseEvent e) {
        if (!gameOver && money >= 25) {
            towers.add(new TDTower(e.getX(), e.getY()));
            money -= 25;
        }
    }

    public void mouseReleased(MouseEvent e) {}
    public void mouseClicked(MouseEvent e) {}
    public void mouseEntered(MouseEvent e) {}
    public void mouseExited(MouseEvent e) {}
}

class TDEnemy {
    double x, y, speed;
    int health, maxHealth;
    int targetIndex = 1;
    boolean finished = false;
    Point[] path;

    TDEnemy(Point[] path, int wave) {
        this.path = path;
        this.x = path[0].x;
        this.y = path[0].y;
        this.health = 3 + wave;
        this.maxHealth = health;
        this.speed = 1.25 + wave * 0.09;
    }

    void move() {
        if (targetIndex >= path.length) {
            finished = true;
            return;
        }
        Point target = path[targetIndex];
        double dx = target.x - x, dy = target.y - y;
        double dist = Math.hypot(dx, dy);
        if (dist < speed) {
            x = target.x; y = target.y; targetIndex++;
        } else {
            x += speed * dx / dist;
            y += speed * dy / dist;
        }
    }

    void draw(Graphics2D g2) {
        g2.setColor(new Color(0,0,0,90));
        g2.fillOval((int)x-15, (int)y+9, 30, 10);
        g2.setColor(new Color(165, 45, 45));
        g2.fillOval((int)x-14, (int)y-14, 28, 28);
        g2.setColor(new Color(255, 160, 160));
        g2.fillOval((int)x-7, (int)y-8, 5, 5);
        g2.fillOval((int)x+3, (int)y-8, 5, 5);

        g2.setColor(Color.BLACK);
        g2.fillRect((int)x-18, (int)y-25, 36, 5);
        g2.setColor(Color.GREEN);
        g2.fillRect((int)x-18, (int)y-25, Math.max(0, 36 * health / maxHealth), 5);
    }
}

class TDTower {
    int x, y, range = 125, cooldown = 0;
    TDTower(int x, int y) { this.x = x; this.y = y; }

    void draw(Graphics2D g2) {
        g2.setColor(new Color(255,255,255,28));
        g2.fillOval(x-range, y-range, range*2, range*2);
        g2.setColor(new Color(30, 40, 85));
        g2.fillOval(x-22, y-22, 44, 44);
        g2.setColor(new Color(75, 110, 210));
        g2.fillOval(x-16, y-16, 32, 32);
        g2.setColor(Color.DARK_GRAY);
        g2.fillRect(x-5, y-34, 10, 32);
        g2.setColor(Color.WHITE);
        g2.drawOval(x-22, y-22, 44, 44);
    }
}

class TDShot {
    double x, y, speed = 9;
    TDEnemy target;
    boolean hit = false;
    TDShot(double x, double y, TDEnemy target) { this.x = x; this.y = y; this.target = target; }

    void move() {
        double dx = target.x - x, dy = target.y - y;
        double dist = Math.hypot(dx, dy);
        if (dist < speed) hit = true;
        else { x += speed * dx / dist; y += speed * dy / dist; }
    }

    void draw(Graphics2D g2) {
        g2.setColor(Color.YELLOW);
        g2.fillOval((int)x-5, (int)y-5, 10, 10);
        g2.setColor(Color.WHITE);
        g2.fillOval((int)x-2, (int)y-2, 4, 4);
    }
}
